src/librustc_mir/build/scope.rs

```diff
@@ -510,7 +545,7 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             resumeblk
         };
 
-        for scope in scopes.iter_mut().filter(|s| !s.drops.is_empty() || s.free.is_some()) {
+        for scope in scopes.iter_mut().filter(|s| s.needs_cleanup) {
```

Comment:
This feels a bit surprising to me. For example, if we have something like:

```rust
{
    let a: Box<i32>; // scope A
    let b: i32; // scope B
    let c: Box<i32>; // scope C
}
```

we will skip over scope B and hence never release the storage for `b`?
    
```